server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -354,6 +371,22 @@ private static Metadata metadataForSnapshot(SnapshotsInProgress.Entry snapshot,
                     builder.put(indexMetadata, false);
                 }
             }
+
+            if (snapshot.userMetadata() != null && snapshot.userMetadata().containsKey(DataStream.DATA_STREAMS_METADATA_FIELD)) {
+                @SuppressWarnings("unchecked")
+                List<String> dataStreamNames = (List<String>) snapshot.userMetadata().get(DataStream.DATA_STREAMS_METADATA_FIELD);
+                Map<String, DataStream> dataStreams = new HashMap<>();
+                for (String dataStreamName : dataStreamNames) {
+                    DataStream dataStream = metadata.dataStreams().get(dataStreamName);
+                    if (dataStream == null) {
+                        assert snapshot.partial() : "Data stream [" + dataStreamName +
```

Comment:
What code guarantees that we don't trip this assertion? It seems to me I can simply remove a data stream while a non-partial snapshot is running right now can't I?

Reply:
Should the delete data stream api then check for snapshots that are in progress and then fail is data stream to be deleted is being snapshotted? Similar to what is done in the delete index api for indices that are being snapshotted?

Reply:
Yea, I'd say so. We treat data streams and indices pretty much the same way in snapshot APIs here so being consistent here makes sense to me.

Reply:
I think that this can be done in a followup?

Reply:
I guess so but this assertion can trip until it's addressed and it's not a great place for the code to be in. We could throw an exception here then in place of the assertion and add a TODO but we are creating a situation where deleting data streams concurrently can either quietly corrupt snapshots or cause them to fail at the very last step of snapshotting (the later option seems ok to me for now).

Reply:
I see, then it is better to address this issue in this PR.

Reply:
I've switched implementation to storing names of data streams in `SnapshotInfo` and added check in delete API for data streams being snapshotted.
I'm not sure if we still need to add check for min node version(both in `SnapshotsService` and `SnapshotInProgress`), as you pointed out we won't be able to create any data stream in mixed cluster so we should never have a situation when we write data that can't be understood by any node.
Please let me know what do you think?